### PR TITLE
fix: added validation to to prevent net total from being negative

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
[Issue](https://bloomstack.com/desk#Form/Issue/ISS-2022-00007)
- [x] Fixes
- [ ] Patches(not required)
- [ ] Test Cases(not required)

<h3>Traceback/Bug Report :</h3>  
When applying a 100% discount to a quote on the grand total it is causing the line items to have a negative value. Felipe took a look at this issue and the problem is related to Net total/Grand total discount.

screenshot-jhaudio com-2022 02 09-22_41_45

<h3>Steps to replicate:</h3> 

* Step 1: Create Quotation
* Step 2: Tax should be applied
* Step 3: Set Additional Discount Percentage to 100 %
* Step 4 : Save Quotation

<h3>Expected Behaviour/Fixes:</h3> 
It does not set total to negative and net rate to negative even if 100 percent discount is applied

screenshot-jhaudio_8008-2022 02 09-22_44_55

### CODE REVIEW

File: erpnext/controllers/taxes_and_totals.py [hunk 1]
```diff
@@ -443,17 +443,28 @@ def apply_discount_amount(self):
...
-							item.precision("net_amount"))
+						if self.doc.discount_amount > self.doc.net_total:
+							item.discount_amount = item.net_amount
+							item.discount_percentage = 100
+							net_total += item.net_amount
+						else:
+							discount_amount_loss = flt(self.doc.net_total - net_total - self.doc.discount_amount,
+								self.doc.precision("net_total"))
+							item.net_amount = flt(item.net_amount + discount_amount_loss,
+								item.precision("net_amount"))
```
Comment (vishdha) [inline erpnext/controllers/taxes_and_totals.py]: Why are doing this change?  as there is check apply_discount_on == Net Total. so your discount amount can not be greater than the Net total. so these changes are not required. 
please validate this and then remove this change

File: erpnext/public/js/controllers/taxes_and_totals.js [hunk 2]
```diff
@@ -538,17 +538,37 @@ erpnext.taxes_and_totals = erpnext.payments.extend({
...
+						item.net_rate = 0;
+						item.amount = 0;
+						item.net_amount = 0;
+						net_total += item.net_amount
+					} else {
+						var discount_amount_loss = flt(me.frm.doc.net_total - net_total
+							- me.frm.doc.discount_amount, precision("net_total"));
+						item.net_amount = flt(item.net_amount + discount_amount_loss,
+							precision("net_amount", item));
+					}
```
Comment (vishdha) [inline erpnext/public/js/controllers/taxes_and_totals.js]: Why are doing this change?  as there is check apply_discount_on == Net Total. so your discount amount can not be greater than the Net total. so these changes are not required.
please validate this and then remove this change

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/controllers/taxes_and_totals.py (hunk 1 above)
Comment (MyuddinKhatri) [reply erpnext/controllers/taxes_and_totals.py]: I did tested after removing this code but it was still setting negative value

File: erpnext/public/js/controllers/taxes_and_totals.js (hunk 2 above)
Comment (MyuddinKhatri) [reply erpnext/public/js/controllers/taxes_and_totals.js]: I did tested after removing this code but it was still setting negative value

**Comment by MyuddinKhatri** [comment]
@vishdha i have done the changes please check.

**Review by neilLasrado — APPROVED** [review]

File: erpnext/controllers/taxes_and_totals.py [hunk 3]
```diff
@@ -443,11 +443,15 @@ def apply_discount_amount(self):
 				else:
 					distributed_amount = flt(self.doc.discount_amount) * item.net_amount
 
-				item.net_amount = flt(item.net_amount - distributed_amount, item.precision("net_amount"))
+				net_amount = flt(item.net_amount - distributed_amount, item.precision("net_amount"))
+				if net_amount < 0 and  not self.doc.is_return:
```
Comment (vishdha) [inline erpnext/controllers/taxes_and_totals.py]:
```suggestion
				if net_amount < 0 and not self.doc.is_return:
```

**Review by vishdha — APPROVED** [review]
Please Add a description in PR why we decided to make net_amount zero. if someone revisits why this change apply

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**